src/backend/distributed/utils/metadata_cache.c

```diff
@@ -964,13 +975,245 @@ CitusHasBeenLoaded(void)
 			 * present during early stages of upgrade operation.
 			 */
 			DistPartitionRelationId();
+
+			/*
+			 * We also set citusCatalogVersion to NULL so that it will be re-read
+			 * in case of extension update.
+			 */
+			citusAvailableVersion = NULL;
+			citusCatalogVersion = NULL;
+		}
+	}
+
+	if (extensionLoaded)
+	{
+		if (!AvailableVersionMatchesBinaryVersion())
+		{
+			ereport(ERROR, (errmsg("citus binaries does not match available extension"),
+							(errdetail("Please restart the server"))));
+		}
+
+		if (!CatalogVersionMatchesBinaryVersion())
+		{
+			ereport(ERROR, (errmsg("citus binaries does not match installed extension"),
+							(errdetail("Please update installed extension with "
+									   "ALTER EXTENSION citus UPDATE;"))));
 		}
 	}
 
 	return extensionLoaded;
 }
 
 
+/*
+ * AvailableVersionMatchesBinaryVersion compares binary version of the Citus and
+ * currently available version from citus.control file. If they are same in major and
+ * minor version numbers , this function returns true. It ignores the patch version.
+ */
+static bool
+AvailableVersionMatchesBinaryVersion()
+{
+	char *availableVersion = CitusAvailableVersion();
+	const char patchSeparator = '-';
+
+	int comparisionLimit = strchr(availableVersion, patchSeparator) - availableVersion;
+
+	if (strncmp(availableVersion, CITUS_BINARY_VERSION, comparisionLimit) == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * AvailableVersionMatchesBinaryVersion compares binary version of the Citus and
+ * catalog version from pg_extemsion catalog table. If they are same in major and
+ * minor version numbers , this function returns true. It ignores the patch version.
+ */
+static bool
+CatalogVersionMatchesBinaryVersion()
+{
+	char *catalogVersion = CitusCatalogVersion();
+	const char patchSeparator = '-';
+
+	int comparisionLimit = strchr(catalogVersion, patchSeparator) - catalogVersion;
+
+	if (strncmp(catalogVersion, CITUS_BINARY_VERSION, comparisionLimit) == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CitusAvailableVersion returns the Citus version from citus.control file. It also
+ * saves the result, thus consecutive calls to CitusExtensionAvailableVersion will
+ * not read the citus.control file again.
```

Comment:
We will not detect it. Problem is citus.control file can change anytime, if we want to detect it we should check it everytime. We cache the result because we call CitusAvailableVersion at almost all queries (more than once generally) and reading citus.control file is not cheap operation.

We discussed this with Ozgun and Andres and they are OK with just detecting version change at the beginning of new connection.